unittests/Interpreter/InterpreterTests.cpp

```diff
@@ -0,0 +1,442 @@
+	x.One.setBit(i);
+	return true;
+      }
+      if (!x.Zero[i] && x.One[i]) {
+	x.Zero.clearBit(i);
+	x.One.clearBit(i);
+	continue;
+      }
+      assert(false && "error in nextKB");
+    }
+    return false;
+  }
+
+  std::string knownBitsString(llvm::KnownBits KB) {
+    std::string S = "";
+    for (int I = 0; I < KB.getBitWidth(); I++) {
+      if (KB.Zero.isNegative())
+	S += "0";
+      else if (KB.One.isNegative())
+	S += "1";
+      else
+	S += "?";
+      KB.Zero <<= 1;
+      KB.One <<= 1;
+    }
+    return S;
+  }
+
+  bool isConcrete(KnownBits x) { return (x.Zero | x.One).isAllOnesValue(); }
+
+  KnownBits merge(KnownBits a, KnownBits b) {
+    assert(a.getBitWidth() == b.getBitWidth() && "Can only merge KnownBits of same width.");
+
+    KnownBits res(a.getBitWidth());
+    for (unsigned i = 0; i < a.getBitWidth(); i++) {
+      if (a.One[i] == b.One[i] && b.One[i])
+	a.One.setBit(i);
+      if (a.Zero[i] == b.Zero[i] && a.Zero[i])
+	a.Zero.setBit(i);
+    }
+    return res;
+  }
+
+  KnownBits setLowest(KnownBits x) {
+    for (int i = 0; i < x.getBitWidth(); i++) {
+      if (!x.Zero[i] && !x.One[i]) {
+	x.One.setBit(i);
+	return x;
+      }
+    }
+    assert(false);
+    return x;
+  }
+
+  KnownBits clearLowest(KnownBits x) {
+    for (int i = 0; i < x.getBitWidth(); i++) {
+      if (!x.Zero[i] && !x.One[i]) {
+	x.Zero.setBit(i);
+	return x;
+      }
+    }
+    assert(false);
+    return x;
+  }
+
+  KnownBits bruteForce(KnownBits x, KnownBits y, Inst::Kind Pred) {
+    if (!isConcrete(x))
+      return merge(bruteForce(setLowest(x), y, Pred),
+		   bruteForce(clearLowest(x), y, Pred));
+    if (!isConcrete(y))
+      return merge(bruteForce(x, setLowest(y), Pred),
+		   bruteForce(x, clearLowest(y), Pred));
+    auto xc = x.getConstant();
+    auto yc = y.getConstant();
+
+    KnownBits res(x.getBitWidth());
+    switch (Pred) {
+    case Inst::AddNUW:
+    case Inst::AddNSW:
+    case Inst::AddNW:
+    case Inst::Add:
+    {
+      auto rc = xc + yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::SubNUW:
+    case Inst::SubNSW:
+    case Inst::SubNW:
+    case Inst::Sub:
+    {
+      auto rc = xc - yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Mul:
+    {
+      auto rc = xc * yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+      break;
+    case Inst::UDiv:
+    {
+      if (yc != 0) {
+	auto rc = xc.udiv(yc);
+	res.One = rc;
+	res.Zero = ~rc;
+      }
+    }
+    break;
+    case Inst::URem:
+    {
+      if (yc != 0) {
+	auto rc = xc.urem(yc);
+	res.One = rc;
+	res.Zero = ~rc;
+      }
+    }
+    break;
+    case Inst::And:
+    {
+      auto rc = xc & yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Or:
+    {
+      auto rc = xc | yc;
+      res.One = rc;
+      res.Zero = ~rc;
+    }
+    break;
+    case Inst::Xor:
+    {
+      auto rc = xc ^ yc;
+      res.One = rc;
+      res.Zero = ~rc;
```

Comment:
the repeated code for these is super ugly, why not just compute the APInt result for all operations and then at the bottom do the res.One and res.Zero thing?